spring-integration-amqp/src/main/java/org/springframework/integration/amqp/inbound/AmqpInboundGateway.java

```diff
@@ -368,38 +386,23 @@ private void process(Message message, org.springframework.messaging.Message<Obje
 					replyTo = AmqpInboundGateway.this.defaultReplyTo;
 				}
 
-				MessagePostProcessor messagePostProcessor =
-						message1 -> {
-							MessageProperties messageProperties = message1.getMessageProperties();
-							String contentEncoding = messageProperties.getContentEncoding();
-							long contentLength = messageProperties.getContentLength();
-							String contentType = messageProperties.getContentType();
-							AmqpInboundGateway.this.headerMapper.fromHeadersToReply(reply.getHeaders(),
-									messageProperties);
-							// clear the replyTo from the original message since we are using it now
-							messageProperties.setReplyTo(null);
-							// reset the content-* properties as determined by the MessageConverter
-							if (StringUtils.hasText(contentEncoding)) {
-								messageProperties.setContentEncoding(contentEncoding);
-							}
-							messageProperties.setContentLength(contentLength);
-							if (contentType != null) {
-								messageProperties.setContentType(contentType);
-							}
-							return message1;
-						};
+				org.springframework.amqp.core.Message amqpMessage =
+						MappingUtils.mapReplyMessage(reply, AmqpInboundGateway.this.amqpMessageConverter,
+								AmqpInboundGateway.this.headerMapper,
+								message.getMessageProperties().getReceivedDeliveryMode(),
+								AmqpInboundGateway.this.replyHeadersMappedLast);
 
 				if (replyTo != null) {
-					AmqpInboundGateway.this.amqpTemplate.convertAndSend(replyTo.getExchangeName(),
-							replyTo.getRoutingKey(), reply.getPayload(), messagePostProcessor);
+					AmqpInboundGateway.this.amqpTemplate.send(replyTo.getExchangeName(), replyTo.getRoutingKey(),
```

Comment:
This is a breaking change - the template might have a different converter to the gateway's converter. See `setMessageConverter()`. 

Perhaps in the private CTOR, extract to template's converter to a different field?